Objects/cown.c

```diff
@@ -0,0 +1,314 @@
+
+static PyObject *PyCown_set_unchecked(PyCownObject *self, PyObject *arg) {
+    // Cowns are cells that hold a reference to a bridge object,
+    // (or another cown or immutable object)
+    const bool is_region_object =
+        Py_IS_TYPE(arg, &PyRegion_Type) && _Py_is_bridge_object(arg);
+    if (is_region_object ||
+        arg->ob_type == &PyCown_Type ||
+        _Py_IsImmutable(arg)) {
+
+        PyObject* old = self->value;
+        Py_XINCREF(arg);
+        self->value = arg;
+
+        // Tell the region that it is owned by a cown,
+        // to enable it to release the cown on close
+        if (is_region_object) {
+            _PyRegion_set_cown_parent(arg, _PyObject_CAST(self));
+            if (_PyRegion_is_closed(arg)) {
+                PyCown_release(self);
+            } else {
+                _Py_atomic_store(&self->state, Cown_PENDING_RELEASE);
+                // TODO: do we need an owning thread for this?
+                PyThreadState *tstate = PyThreadState_Get();
+                self->owning_thread = tstate->thread_id;
+            }
+        } else {
+            // We can release this cown immediately
+            // _Py_atomic_store(&self->state, Cown_RELEASED);
+            // self->owning_thread = 0;
+            PyCown_release(self);
+        }
+
+        return old ? old : Py_None;
+    } else {
+        // Invalid cown content
+        PyErr_SetString(PyExc_RuntimeError,
+            "Cowns can only store bridge objects, immutable objects or other cowns!");
+        return NULL;
+    }
+}
+
+static PyObject *PyCown_set(PyCownObject *self, PyObject *arg) {
+    BAIL_UNLESS_ACQUIRED(self, "Attempt to set value of unacquired cown");
+    return PyCown_set_unchecked(self, arg);
+}
+
+static int PyCown_clear(PyCownObject *self) {
+    Py_CLEAR(self->value);
+    return 0;
+}
+
+static PyObject *PyCown_repr(PyCownObject *self) {
+#ifdef PYDEBUG
+    if (STATE(self) == Cown_ACQUIRED) {
+        return PyUnicode_FromFormat(
+            "Cown(status=acquired by thread %zd,value=%S)",
+            PyThreadState_Get()->thread_id,
+            PyObject_Repr(self->value)
+        );
+    } else {
+        return PyUnicode_FromFormat(
+            "Cown(status=%s,value=%S)",
+            STATE(self) == Cown_RELEASED
+                ? "released"
+                : "pending-release",
+            PyObject_Repr(self->value)
+        );
+    }
+#else
+    if (STATE(self) == Cown_ACQUIRED) {
+        return PyUnicode_FromFormat(
+            "Cown(status=acquired by thread %zd)",
+            PyThreadState_Get()->thread_id
+        );
+    } else {
+        return PyUnicode_FromFormat(
+            "Cown(status=%s)",
+            STATE(self) == Cown_RELEASED
+                ? "released"
+                : "pending-release"
+        );
+    }
+#endif
+}
+
+// Define the CownType with methods
+static PyMethodDef PyCown_methods[] = {
+    {"acquire", (PyCFunction)PyCown_acquire, METH_NOARGS, "Acquire the cown."},
```

Comment:
This triggers several warnings in the GH CI. Can you take a look at it? Not sure how we resolved this for regions.

Reply:
I'll add a zero arity C function pointer type and use that instead.

Reply:
OK, that didn't work. I'll revert a change requested by Matt P (why do the cown functions have unused arguments?) and put the arguments back but document why instead.